File: php/src/RankCache.php
<?php

declare(strict_types=1);

namespace TurboToken;

class RankCache
{
    public static function getCacheDir(): string
    {
        $env = getenv('TURBOTOKEN_CACHE_DIR');
        if ($env !== false && $env !== '') {
            return $env;
        }

        $home = getenv('HOME') ?: (getenv('USERPROFILE') ?: sys_get_temp_dir());
        return $home . DIRECTORY_SEPARATOR . '.cache' . DIRECTORY_SEPARATOR . 'turbotoken';
    }

    public static function ensureRankFile(string $name): string
    {
        $spec = Registry::getEncodingSpec($name);
        $cacheDir = self::getCacheDir();

        if (!is_dir($cacheDir)) {
            if (!mkdir($cacheDir, 0755, true) && !is_dir($cacheDir)) {
                throw new TurboTokenException("Failed to create cache directory: $cacheDir");
            }
        }

        $filePath = $cacheDir . DIRECTORY_SEPARATOR . $spec->name . '.tiktoken';

        if (file_exists($filePath)) {
            return $filePath;
        }

        $context = stream_context_create([
            'http' => [
                'timeout' => 60,
                'user_agent' => 'turbotoken-php/0.1.0',
            ],
        ]);

        $data = @file_get_contents($spec->rankFileUrl, false, $context);
        if ($data === false) {
            throw new TurboTokenException("Failed to download rank file from: {$spec->rankFileUrl}");
        }

        if (file_put_contents($filePath, $data) === false) {
            throw new TurboTokenException("Failed to write rank file to: $filePath");
        }

        return $filePath;
    }

    public static function readRankFile(string $name): string
    {
        $filePath = self::ensureRankFile($name);
        $data = file_get_contents($filePath);
        if ($data === false) {
            throw new TurboTokenException("Failed to read rank file: $filePath");
        }
        return $data;
    }
}

File: wrappers/lua/src/turbotoken/rank_cache.lua
local registry = require("turbotoken.registry")

local M = {}

function M.cache_dir()
    local env = os.getenv("TURBOTOKEN_CACHE_DIR")
    if env and env ~= "" then
        return env
    end

    local home = os.getenv("HOME") or os.getenv("USERPROFILE") or os.tmpname():match("(.+)/")
    return home .. "/.cache/turbotoken"
end

local function file_exists(path)
    local f = io.open(path, "rb")
    if f then
        f:close()
        return true
    end
    return false
end

local function mkdir_p(path)
    -- Use os.execute for cross-platform directory creation
    if package.config:sub(1, 1) == "\\" then
        os.execute('mkdir "' .. path .. '" 2>NUL')
    else
        os.execute('mkdir -p "' .. path .. '"')
    end
end

local function download_file(url, dest)
    -- Try LuaSocket first
    local ok, http = pcall(require, "socket.http")
    if ok then
        local ltn12 = require("ltn12")
        local sink = ltn12.sink.file(io.open(dest, "wb"))
        local result, status = http.request({
            url = url,
            sink = sink,
        })
        if result and status == 200 then
            return true
        end
        os.remove(dest)
    end

    -- Fallback to curl
    local cmd = string.format('curl -sL -o "%s" "%s"', dest, url)
    local ret = os.execute(cmd)
    if ret == 0 or ret == true then
        return true
    end

    -- Fallback to wget
    cmd = string.format('wget -q -O "%s" "%s"', dest, url)
    ret = os.execute(cmd)
    if ret == 0 or ret == true then
        return true
    end

    error("Failed to download rank file from: " .. url)
end

function M.ensure_rank_file(name)
    local spec = registry.get_encoding_spec(name)
    local dir = M.cache_dir()
    local file_path = dir .. "/" .. spec.name .. ".tiktoken"

    if file_exists(file_path) then
        return file_path
    end

    mkdir_p(dir)
    download_file(spec.rank_file_url, file_path)

    if not file_exists(file_path) then
        error("Failed to create rank file: " .. file_path)
    end

    return file_path
end

function M.read_rank_file(name)
    local file_path = M.ensure_rank_file(name)
    local f = io.open(file_path, "rb")
    if not f then
        error("Failed to read rank file: " .. file_path)
    end
    local data = f:read("*a")
    f:close()
    return data
end

return M
